[Doc] Customize KubeRay container commands (#41651)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/pod-command.md

(kuberay-pod-command)=

# Specify container commands for Ray head/worker Pods

> KubeRay generates a `ray start` command for each Ray Pod.

Reviewer: It may be helpful to include an example of the full command here, just so users know what it looks like

Sometimes, you may want to execute certain commands either before or after the ray start command, or you may wish to define the container's command yourself.
This document shows you how to do that.

> ## Part 1: Specify your custom container command without any generated `ray start` command

Reviewer: ```suggestion
## Part 1: Specify a custom container command, optionally including the generated `ray start` command
```
The original title is a bit confusing because in practice, the custom container command will usually have the generated `ray start` command in it, so it's not "without" the generated command

> Starting with KubeRay v1.1.0, if users add the annotation `ray.io/overwrite-container-cmd: "true"` to a RayCluster, KubeRay respects the container `command` and `args` as provided by the users, without including any generated command.

Reviewer: ```suggestion
Starting with KubeRay v1.1.0, if users add the annotation `ray.io/overwrite-container-cmd: "true"` to a RayCluster, KubeRay respects the container `command` and `args` as provided by the users, without including the generated `ray start` command.
```
Nit. Just to be a bit more explicit.

Reviewer: It would also be helpful to include the definition/explanation of `KUBERAY_GEN_RAY_START_CMD` here, in the text, instead of just in the code comment (users might not expect to see a key definition in the code comment instead of the main text)

Author: This is not precise enough. The generated commands include `ulimit` and `ray start` commands. I add more details about both of them.

```yaml
apiVersion: ray.io/v1
kind: RayCluster
metadata:
  annotations:
    # If this annotation is set to "true", KubeRay will respect the container `command` and `args`.
    ray.io/overwrite-container-cmd: "true"
  ...
spec:
  headGroupSpec:
    rayStartParams: {}
    # Pod template
    template:
      spec:
        containers:
        - name: ray-head
          image: rayproject/ray:2.8.0
          # Because the annotation "ray.io/overwrite-container-cmd" is set to "true",
          # KubeRay will overwrite the generated container command with `command` and
          # `args` in the following. Hence, you need to specify the `ulimit` command

>           # by yourself to avoid Ray scalability issues.

Reviewer: Reading this, the reader will be curious: what happens if the annotation is not set or set to false? I guess `args` and `command` will be ignored, hopefully with a warning?

Author: > I guess args and command will be ignored, hopefully with a warning?

No, the default behavior is "Part 2".

          command: ["/bin/bash", "-lc", "--"]
          # Starting from v1.1.0, KubeRay injects the environment variable `KUBERAY_GEN_RAY_START_CMD`
          # into the Ray container. This variable can be used to retrieve the generated Ray start command.
          # Note that this environment variable does not include the `ulimit` command.
          args: ["echo head; ulimit -n 65536; $KUBERAY_GEN_RAY_START_CMD"]
          ...
```

The preceding example YAML is a part of [ray-cluster.overwrite-command.yaml](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray-cluster.overwrite-command.yaml).

* `metadata.annotations.ray.io/overwrite-container-cmd: "true"`: This annotation tells KubeRay to respect the container `command` and `args` as provided by the users, without including any generated command.

> * `ulimit -n 65536`: This command is necessary to avoid Ray scalability issues caused by running out of file descriptors.

Reviewer: I forgot (and the reader might be curious), why didn't we just make ulimit -n 65536 part of the KUBERAY_GEN_RAY_START_CMD? Is it because some users might want to delete the ulimit part?

Reviewer: If there's an interesting explanation for the above question it might be helpful to write it in the doc, maybe here or at the line below where we say "Note that this environment variable doesn't include the `ulimit` command."

Author: > why didn't we just make ulimit -n 65536 part of the KUBERAY_GEN_RAY_START_CMD? Is it because some users might want to delete the ulimit part?

Some users want to run some custom commands between `ulimit` and `ray start`. https://github.com/ray-project/kuberay/issues/1560#issuecomment-1836965826 is an example.

Reviewer: Ah right, I forgot. Makes sense

If you don't set the annotation, KubeRay automatically injects the `ulimit` command into the container.

* `$KUBERAY_GEN_RAY_START_CMD`: Starting from KubeRay v1.1.0, KubeRay injects the environment variable `KUBERAY_GEN_RAY_START_CMD` into the Ray container for both head and worker Pods to store the `ray start` command generated by KubeRay.
Note that this environment variable doesn't include the `ulimit` command.

The head Pod's `command`/`args` looks like the following:

```yaml
Command:
  /bin/bash
  -lc
  --
Args:
  echo head; ulimit -n 65536; $KUBERAY_GEN_RAY_START_CMD
```

> ## Part 2: Execute commands before the generated `ray start` command

Reviewer: Is there any advantage to this approach actually, or should the user always just use Part 1?

Author: > Is there any advantage to this approach actually, or should the user always just use Part 1?

No, this is just for the compatibility.

Author: I believed some users use this, and some KubeRay examples also use this.

> If you only want to execute commands before the generated command, you don't need to set the annotation `ray.io/overwrite-container-cmd: "true"`.

Reviewer: This part is a bit confusing. In the example above for Part 1, didn't we also only want to execute commands (`echo hello`) before the generated command? If that's the case, we should choose a different example for Part 1

Author: https://github.com/ray-project/kuberay/pull/1722

    # Args:
    #    echo 123  456  && ulimit -n 65536; ray start --head  --dashboard-host=0.0.0.0  --num-cpus=1  --block  --metrics-export-port=8080  --memory=2147483648
    ```

> ## Timing 2: After `ray start` (RayCluster is ready)

Reviewer: I had to think a bit to remember that the reason we don't need this section anymore is because we can use Part 1 to just directly add post-start commands after the KUBERAY_GEN_RAY_START_COMMAND.  

I think it would help to make this really obvious and explicit in the doc for part 1, maybe just with a quick sentence or example. Or maybe just put a post start command in the existing example.

Author: This is an over-explanation for me. I will update it only if you feel strongly about its necessity.
